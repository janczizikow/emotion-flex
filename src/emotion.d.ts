import {} from '@emotion/react/types/css-prop';
import { Theme as EmotionFlexTheme } from './theme'

declare module '@emotion/react' {
  export interface Theme extends EmotionFlexTheme {}
}