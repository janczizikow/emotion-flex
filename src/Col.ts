import facepaint from 'facepaint';
import styled from '@emotion/styled';
import { isInteger, getTheme, getBreakpoints } from './utils';

type TwoToEleven = 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
type OneToTwelve = 1 | TwoToEleven | 12;
type ZeroToTwelve = 0 | OneToTwelve;

type Offset = 0 | 1 | TwoToEleven;
type Order = ZeroToTwelve | 'first' | 'last';
type Column = OneToTwelve;

export interface ColProps {
  /**
   * Removes the padding around Column
   */
  noGutters?: boolean;
  /**
   * The number of columns to offset on extra small devices
   */
  xsOffset?: Offset;
  xsOrder?: Order;
  /**
   * The number of columns to span on extra small devices
   */
  xs?: Column;
  /**
   * The number of columns to offset on small devices
   */
  smOffset?: Offset;
  smOrder?: Order;
  /**
   * The number of columns to span on small devices
   */
  sm?: Column;
  /**
   * The number of columns to offset on medium devices
   */
  mdOffset?: Offset;
  mdOrder?: Order;
  /**
   * The number of columns to span on medium devices
   */
  md?: Column;
  /**
   * The number of columns to offset on large devices
   */
  lgOffset?: Offset;
  lgOrder?: Order;
  /**
   * The number of columns to span on large devices
   */
  lg?: Column;
  /**
   * The number of columns to offset on extra large devices
   */
  xlOffset?: Offset;
  xlOrder?: Order;
  /**
   * The number of columns to span on extra large devices
   */
  xl?: Column;
}

const toOrder = (order?: Order) => {
  if (!order || typeof order === 'number') {
    return order;
  }
  return order === 'first' ? -1 : 13;
};

const toPercent = (n: number) => {
  const percentage = (n / 12) * 100;
  return isInteger(percentage) ? `${percentage}%` : `${percentage.toFixed(6)}%`;
};

export const Col = styled.div<ColProps>`
  position: relative;
  width: 100%;
  padding-right: ${({ noGutters, theme }) =>
    noGutters ? 0 : getTheme(theme).gutters}px;
  padding-left: ${({ noGutters, theme }) =>
    noGutters ? 0 : getTheme(theme).gutters}px;

  ${p =>
    facepaint(getBreakpoints(getTheme(p.theme)))({
      marginLeft: [
        p.xsOffset && toPercent(p.xsOffset),
        p.smOffset && toPercent(p.smOffset),
        p.mdOffset && toPercent(p.mdOffset),
        p.lgOffset && toPercent(p.lgOffset),
        p.xlOffset && toPercent(p.xlOffset),
      ],
      flex: [
        p.xs ? `1 0 ${toPercent(p.xs)}` : '1 0 0',
        p.sm && `1 0 ${toPercent(p.sm)}`,
        p.md && `1 0 ${toPercent(p.md)}`,
        p.lg && `1 0 ${toPercent(p.lg)}`,
        p.xl && `1 0 ${toPercent(p.xl)}`,
      ],
      maxWidth: [
        p.xs ? toPercent(p.xs) : '100%',
        p.sm && toPercent(p.sm),
        p.md && toPercent(p.md),
        p.lg && toPercent(p.lg),
        p.xl && toPercent(p.xl),
      ],
      order: [
        toOrder(p.xsOrder),
        toOrder(p.smOrder),
        toOrder(p.mdOrder),
        toOrder(p.lgOrder),
        toOrder(p.xlOrder),
      ],
    })}
`;

Col.displayName = 'Col';
