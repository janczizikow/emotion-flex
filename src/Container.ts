import facepaint from 'facepaint';
import styled from '@emotion/styled';
import { getTheme, getBreakpoints } from './utils';

export interface ContainerProps {
  /**
   * Allow the Container to fill all of it's availble horizontal space.
   */
  fluid?: boolean;
}

export const Container = styled.div<ContainerProps>`
  padding-right: ${({ theme }) => getTheme(theme).gutters}px;
  padding-left: ${({ theme }) => getTheme(theme).gutters}px;
  margin-right: auto;
  margin-left: auto;

  ${({ fluid, theme }) =>
    fluid
      ? { maxWidth: '100%' }
      : facepaint(getBreakpoints(getTheme(theme)))({
          maxWidth: [
            '100%',
            getTheme(theme).maxWidths.sm,
            getTheme(theme).maxWidths.md,
            getTheme(theme).maxWidths.lg,
            getTheme(theme).maxWidths.xl,
          ],
        })}
`;

Container.displayName = 'Container';
