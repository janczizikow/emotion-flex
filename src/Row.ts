import styled from '@emotion/styled';
import { getTheme } from './utils';

export interface RowProps {
  /**
   * Removes Row's negative margins.
   */
  noGutters?: boolean;
}

export const Row = styled.div<RowProps>`
  display: flex;
  flex-wrap: wrap;
  margin-left: ${({ noGutters, theme }) =>
    noGutters ? 0 : `-${getTheme(theme).gutters}px`};
  margin-right: ${({ noGutters, theme }) =>
    noGutters ? 0 : `-${getTheme(theme).gutters}px`};
`;

Row.displayName = 'Row';
