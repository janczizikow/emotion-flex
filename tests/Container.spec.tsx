import React from 'react';
import { create } from 'react-test-renderer';
import { ThemeProvider } from '@emotion/react';
import { customTheme } from './test-helpers';

import { Container } from '../src';

describe('<Container />', () => {
  it('renders with default styles', () => {
    const tree = create(<Container />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with default fluid styles', () => {
    const tree = create(<Container fluid />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with custom styles when given a custom theme', () => {
    const tree = create(
      <ThemeProvider theme={customTheme}>
        <Container />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
