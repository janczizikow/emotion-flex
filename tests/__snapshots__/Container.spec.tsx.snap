// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> renders with custom styles when given a custom theme 1`] = `
.emotion-0 {
  padding-right: 16px;
  padding-left: 16px;
  margin-right: auto;
  margin-left: auto;
  max-width: 100%;
}

@media (min-width: 600px) {
  .emotion-0 {
    max-width: 576px;
  }
}

@media (min-width: 960px) {
  .emotion-0 {
    max-width: 768px;
  }
}

@media (min-width: 1280px) {
  .emotion-0 {
    max-width: 992px;
  }
}

@media (min-width: 1920px) {
  .emotion-0 {
    max-width: 1200px;
  }
}

<div
  className="emotion-0"
/>
`;

exports[`<Container /> renders with default fluid styles 1`] = `
.emotion-0 {
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
  max-width: 100%;
}

<div
  className="emotion-0"
/>
`;

exports[`<Container /> renders with default styles 1`] = `
.emotion-0 {
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
  max-width: 100%;
}

@media (min-width: 576px) {
  .emotion-0 {
    max-width: 540px;
  }
}

@media (min-width: 768px) {
  .emotion-0 {
    max-width: 720px;
  }
}

@media (min-width: 992px) {
  .emotion-0 {
    max-width: 960px;
  }
}

@media (min-width: 1200px) {
  .emotion-0 {
    max-width: 1140px;
  }
}

<div
  className="emotion-0"
/>
`;
